[BXMSDOC-8419] Updating messaging addon documentation (#4165)
kiegroup/kie-docs: doc-content/kogito-docs/src/main/asciidoc/creating-running/chap-kogito-creating-running.adoc

* xref:con-data-index-service_kogito-configuring[{PRODUCT} Data Index Service]
* xref:con-trusty-service_kogito-configuring[{PRODUCT} Trusty Service and Explainability Service]

> To configure messaging capabilities for your Kogito services, you should add the messaging add-on as a dependency in the pom.xml file of your Kogito project:

Reviewer: ```suggestion
To configure messaging capabilities for your Kogito services, you must add the messaging add-on as a dependency in the `pom.xml` file of your Kogito project:
```

> To configure messaging capabilities for your Kogito services, you should add the messaging add-on as a dependency in the pom.xml file of your Kogito project:

Reviewer 2: [style]

Add formatting for file:

s/pom.xml/`pom.xml`

It appears the dependency is required, so:

s/you should add/you need to add/you must add

</dependency>
----

For Quarkus, the {PRODUCT} process event add-on implementation is based on https://smallrye.io/smallrye-reactive-messaging/smallrye-reactive-messaging/2/connectors/connectors.html[Smallrye Messaging] library, which provides a set of connectors for event brokers, such as JMS, AMQP, and Kafka. Therefore, the {PRODUCT} process event add-on is not specifically combined with any event broker, but also requires additional configuration to use a suitable Smallrye connector.

> For example, to use Kafka as an event broker in Quarkus, you should add the following dependency in the `pom.xml` file of your {PRODUCT} project:

Reviewer: ```suggestion
For example, to use Kafka as an event broker in Quarkus, you must add the following dependency in the `pom.xml` file of your {PRODUCT} project:
```

</dependency>
----

> Smallrye uses an abstraction named channel, which real implementation depends on the underlying event broker. For example, when using Kakfa, every channel is mapped to a topic. 

Reviewer: ```suggestion
The Smallrye uses an abstraction named channel, where the real implementation of such a channel depends on the underlying event broker. For example, when using Kafka, every channel is mapped to a topic. 
```

> Channels are configured through properties. The pattern for these properties is `mp.messaging.[incoming|outgoing].<channel name>.<property_name>`.

Reviewer 2: [translation question] "through' may be misunderstood in translation. Would "by using" be better?

> Kogito allows different channel mapping strategies:

Reviewer 3: ```suggestion
{PRODUCT} allows different channel mapping strategies:
```

* Define one default incoming channel to receive all the incoming messages and one default outgoing channel to store all the published messages.
* Define a channel for each message name so that every message type has a dedicated channel.

> * Define a channel for certain message names and let the non mapped message names to use the default incoming or outgoing channel. 

Author: changed to not mapped

The name for default incoming channel is `kogito_incoming_stream` and for default outgoing channel is `kogito_outgoing_stream`. 

You can change the default incoming and outgoing topic name as follows:

.Quarkus
[source]
----
kogito.addon.messaging.incoming.defaultName=<default channel name>
kogito.addon.messaging.outgoing.defaultName=<default channel name>
----

> .Springboot

Reviewer 3: ```suggestion
.Spring Boot
```

[source]
----
kogito.addon.cloudevents.kafka.kogito_incoming_stream=<default channel name>
kogito.addon.cloudevents.kafka.kogito_outgoing_stream=<default channel name>
----

> Optionally, for Kafka connector, you can define the Kafka topic to be used for that channel using the following form:

Author: Kafka is the name of a tool, so it is singular and uppercase.

`mp.messaging.[incoming|outgoing].<channel name>.topic = <topic name>`

NOTE: If topic property is not found, then the topic name is considered to be the same as Kafka name.

> ==== Providing a key for a kafka record

Reviewer 3: ```suggestion
==== Providing a key for a Kafka record
```

> In order to specify the key to your kafka record, you can use a custom https://github.com/kiegroup/kogito-runtimes/blob/main/quarkus/addons/messaging/common/src/main/java/org/kie/kogito/addon/quarkus/messaging/common/message/MessageDecorator.java[MessageDecorator] implementation. 

Reviewer 3: ```suggestion
To specify the key to your Kafka record, you can use a custom https://github.com/kiegroup/kogito-runtimes/blob/main/quarkus/addons/messaging/common/src/main/java/org/kie/kogito/addon/quarkus/messaging/common/message/MessageDecorator.java[MessageDecorator] implementation. 
```

You can find an example https://github.com/kiegroup/kogito-examples/blob/main/kogito-quarkus-examples/process-kafka-avro-multi-quarkus/src/main/java/org/acme/travel/StringKeyDecorator.java[here]

==== Customizing message format

> By default, Kogito uses Cloud event json format for messaging. You can change the default format by providing your own https://github.com/kiegroup/kogito-runtimes/blob/main/api/kogito-events-api/src/main/java/org/kie/kogito/event/EventMarshaller.java[event marshaler] implementation. 

Reviewer 3: ```suggestion
By default, Kogito uses Cloud Event JSON format for messaging. However, you can change the default format by providing your own https://github.com/kiegroup/kogito-runtimes/blob/main/api/kogito-events-api/src/main/java/org/kie/kogito/event/EventMarshaller.java[event marshaler] implementation. 
```

> You can find an example https://github.com/kiegroup/kogito-examples/blob/main/kogito-quarkus-examples/process-kafka-avro-multi-quarkus/src/main/java/org/acme/travel/AvroEventMarshaller.java[here]

Reviewer 2: [style] [Link] or [Here] is not used, replace this with the name of the example, and add the general location:

For more information, see `link`[Name of topic] in the _Name of doc_ guide. For example:

https://github.com/kiegroup/kogito-examples/blob/main/kogito-quarkus-examples/process-kafka-avro-multi-quarkus/src/main/java/org/acme/travel/AvroEventMarshaller.java[AvroEventMarshaller.java example] in the https://github.com/kiegroup/kogito-examples[Kogito-examples repository].

[style question] Can this be rephrased to more closely resemble our usual linking phrase? For example:

For an example of an event marshaller, see https://github.com/kiegroup/kogito-examples/blob/main/kogito-quarkus-examples/process-kafka-avro-multi-quarkus/src/main/java/org/acme/travel/AvroEventMarshaller.java[AvroEventMarshaller.java example] in the https://github.com/kiegroup/kogito-examples[Kogito-examples repository].

Author: Agree with not using here, but why we should add the general repository if it is already included in the example link. It does not really make any sense to me.

</dependency>
----

> Smallrye uses an abstraction named channel, which real implementation depends on the underlying event broker. For example, when using Kakfa, every channel is mapped to a topic. 

Reviewer 2: [style]

Does Smallrey need the article?

Should "channel" be formatted as an element?

Regarding clauses. While "which" is a nonrestrictive clause and "that" is a restrictive clause, I think this needs to be a relative clause "of which" or the channel needs to be referred to specifically. For example:

"Smallrye uses an abstraction named channel, of which the implementation depends on the underlying event broker."

Alternatively, this sentence can be separated to make it easier to understand. For example:

"Smallrye uses an abstraction named "channel". The implementation of channels depend on the underlying event broker."

s/Kakfa/Kafka

> Smallrye uses an abstraction named channel, which real implementation depends on the underlying event broker. For example, when using Kakfa, every channel is mapped to a topic. 

Reviewer 4: maybe worth mentioning, different channels can be mapped to a single topic not just a 1:1 relationship, this could be useful for some use cases.

Author: I think this is implicit latter. And given that every sentence arise a lot of style questions, I prefer to be brief
